internal/worker/concurrency_auto_scaler.go

```diff
@@ -0,0 +1,305 @@
+		scope        tally.Scope
+		clock        clockwork.Clock
+
+		concurrency *ConcurrencyLimit
+		cooldown    time.Duration
+		updateTick  time.Duration
+
+		// state of autoscaler
+		lock    sync.RWMutex
+		enabled bool
+
+		// poller
+		pollerInitCount        int
+		pollerMaxCount         int
+		pollerMinCount         int
+		pollerWaitTime         *rollingAverage[time.Duration]
+		pollerPermitLastUpdate time.Time
+	}
+
+	ConcurrencyAutoScalerInput struct {
+		Concurrency    *ConcurrencyLimit
+		Cooldown       time.Duration // cooldown time of update
+		Tick           time.Duration // frequency of update check
+		PollerMaxCount int
+		PollerMinCount int
+		Logger         *zap.Logger
+		Scope          tally.Scope
+		Clock          clockwork.Clock
+	}
+
+	autoScalerEvent string
+)
+
+func NewConcurrencyAutoScaler(input ConcurrencyAutoScalerInput) *ConcurrencyAutoScaler {
+	tick := defaultAutoScalerUpdateTick
+	if input.Tick != 0 {
+		tick = input.Tick
+	}
+	return &ConcurrencyAutoScaler{
+		shutdownChan:           make(chan struct{}),
+		concurrency:            input.Concurrency,
+		cooldown:               input.Cooldown,
+		log:                    input.Logger.Named(metrics.ConcurrencyAutoScalerScope),
+		scope:                  input.Scope.SubScope(metrics.ConcurrencyAutoScalerScope),
+		clock:                  input.Clock,
+		updateTick:             tick,
+		enabled:                false, // initial value should be false and is only turned on from auto config hint
+		pollerInitCount:        input.Concurrency.PollerPermit.Quota(),
+		pollerMaxCount:         input.PollerMaxCount,
+		pollerMinCount:         input.PollerMinCount,
+		pollerWaitTime:         newRollingAverage[time.Duration](numberOfPollsInRollingAverage),
+		pollerPermitLastUpdate: input.Clock.Now(),
+	}
+}
+
+func (c *ConcurrencyAutoScaler) Start() {
+	c.logEvent(autoScalerEventStart)
+
+	c.wg.Add(1)
+
+	go func() {
+		defer c.wg.Done()
+		ticker := c.clock.NewTicker(c.updateTick)
+		defer ticker.Stop()
+		for {
+			select {
+			case <-c.shutdownChan:
+				return
+			case <-ticker.Chan():
+				c.lock.Lock()
+				c.logEvent(autoScalerEventMetrics)
+				c.updatePollerPermit()
+				c.lock.Unlock()
+			}
+		}
+	}()
+}
+
+func (c *ConcurrencyAutoScaler) Stop() {
+	close(c.shutdownChan)
+	c.wg.Wait()
+	c.logEvent(autoScalerEventStop)
+}
+
+// ProcessPollerHint reads the poller response hint and take actions in a transactional way
+// 1. update poller wait time
+// 2. enable/disable auto scaler
+func (c *ConcurrencyAutoScaler) ProcessPollerHint(hint *shared.AutoConfigHint) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if hint == nil {
+		return
+	}
+	if hint.PollerWaitTimeInMs != nil {
+		waitTimeInMs := *hint.PollerWaitTimeInMs
+		c.pollerWaitTime.Add(time.Millisecond * time.Duration(waitTimeInMs))
+	}
+
+	var shouldEnable bool
+	if hint.EnableAutoConfig != nil && *hint.EnableAutoConfig {
+		shouldEnable = true
+	}
+	if shouldEnable != c.enabled { // flag switched
```

Comment:
nit: consider an early abort?

```
if hint.EnableAutoConfig == nil || !*hint.EnableAutoConfig {
   return 
}
``

Reply:
Actually, we need a resetConcurrency() action here if the pollers change from enabled to disabled state to avoid locking them in a unhealthy state. 

For example, when users found an issue and wish to turn off auto scaler, it's best to reset concurrency limit to original value.